subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -237,40 +260,65 @@ public FileCollection fileCollection(Dependency... dependencies) {
         return new ConfigurationFileCollection(WrapUtil.toLinkedSet(dependencies));
     }
 
-    public void includedInResolveResult() {
-        includedInResult = true;
+    public void markAsObserved() {
+        synchronized (lock) {
+            if (state == InternalState.UNOBSERVED) {
+                state = InternalState.OBSERVED;
+            }
+        }
         for (Configuration configuration : extendsFrom) {
-            ((ConfigurationInternal) configuration).includedInResolveResult();
+            ((ConfigurationInternal) configuration).markAsObserved();
         }
     }
 
     public ResolvedConfiguration getResolvedConfiguration() {
-        resolveNow();
+        resolveNow(InternalState.RESULTS_RESOLVED);
         return cachedResolverResults.getResolvedConfiguration();
     }
 
-    private void resolveNow() {
+    private void resolveNow(InternalState requestedState) {
         synchronized (lock) {
-            if (state == State.UNRESOLVED) {
+            boolean needsResolve = state == InternalState.UNOBSERVED || state == InternalState.OBSERVED || modified;
```

Comment:
This isn't quite right. We should never re-resolve the configuration if it is in the RESULTS_RESOLVED state, even if it has been mutated. This means that the deprecation message on these changes should mention that the changes will be ignored.